Mathlib/Analysis/Convex/Side.lean

```diff
@@ -829,3 +829,219 @@ theorem isPreconnected_setOf_sOppSide (s : AffineSubspace ℝ P) (x : P) :
+      (Finset.mem_univ j), Finset.affineCombination_vsub]
+    let w₃ : Fin (n + 1) → R :=
+      w₂ - w₂ i • (w₁ i)⁻¹ • (w₁ - Finset.affineCombinationSingleWeights R j)
+    have hw₃1 : ∑ k, w₃ k = 1 := by simp [w₃, hw₂, ← Finset.mul_sum, hw₁]
+    have hw₃i : w₃ i = 0 := by simp [w₃, hj.symm, h0]
+    refine ⟨Finset.univ.affineCombination R s.points w₃, ?_, ?_⟩
+    · rw [range_faceOpposite_points]
+      exact affineCombination_mem_affineSpan_image hw₃1 (by simpa using hw₃i) s.points
+    · simp only [w₃, Finset.affineCombination_vsub, sub_sub_cancel, smul_smul, map_smul,
+        sameRay_smul_right_iff]
+      left
+      rcases h0.lt_or_gt with h | h
+      · rw [sign_neg h, eq_comm, sign_eq_neg_one_iff] at hs
+        exact (mul_pos_of_neg_of_neg hs (inv_neg''.2 h)).le
+      · rw [sign_pos h, eq_comm, sign_eq_one_iff] at hs
+        positivity
+  · intro h
+    rw [range_faceOpposite_points] at h
+    exact h0 (s.independent.eq_zero_of_affineCombination_mem_affineSpan hw₁ h (Finset.mem_univ i)
+      (by simp))
+  · intro h
+    rw [range_faceOpposite_points] at h
+    exact h0' (s.independent.eq_zero_of_affineCombination_mem_affineSpan hw₂ h (Finset.mem_univ i)
+      (by simp))
+
+lemma sOppSide_affineSpan_faceOpposite_of_pos_of_neg {w₁ w₂ : Fin (n + 1) → R}
+    (hw₁ : ∑ j, w₁ j = 1) (hw₂ : ∑ j, w₂ j = 1) {i : Fin (n + 1)} (hs₁ : 0 < w₁ i)
+    (hs₂ : w₂ i < 0) :
+    (affineSpan R (Set.range (s.faceOpposite i).points)).SOppSide
+      (Finset.univ.affineCombination R s.points w₁)
+      (Finset.univ.affineCombination R s.points w₂) := by
+  let w₃ : Fin (n + 1) → R := lineMap w₁ w₂ (w₁ i / (w₁ i - w₂ i))
+  have hp : 0 < w₁ i - w₂ i := by grind
+  have hw₃ : ∑ j, w₃ j = 1 := by
+    simp [w₃, lineMap_apply, Finset.sum_add_distrib, ← Finset.mul_sum, hw₁, hw₂]
+  have h : Sbtw R w₁ w₃ w₂ := sbtw_lineMap_iff.2
+    ⟨(by grind), div_pos hs₁ hp, (div_lt_one hp).2 (by grind)⟩
+  have h' : Sbtw R (Finset.univ.affineCombination R s.points w₁)
+      (Finset.univ.affineCombination R s.points w₃)
+      (Finset.univ.affineCombination R s.points w₂) := by
+    refine ⟨h.wbtw.map _, fun he ↦ ?_, fun he ↦ ?_⟩
+    · replace he := (affineIndependent_iff_eq_of_fintype_affineCombination_eq R s.points).1
+        s.independent w₃ w₁ hw₃ hw₁ he
+      simp_rw [← he, w₃] at hs₁
+      simp only [lineMap_apply, vsub_eq_sub, Pi.vadd_apply', Pi.smul_apply, Pi.sub_apply,
+        smul_eq_mul, vadd_eq_add] at hs₁
+      rw [← neg_sub (w₁ i) (w₂ i), mul_neg, div_mul_cancel₀ _ hp.ne'] at hs₁
+      simp at hs₁
+    · replace he := (affineIndependent_iff_eq_of_fintype_affineCombination_eq R s.points).1
+        s.independent w₃ w₂ hw₃ hw₂ he
+      simp_rw [← he, w₃] at hs₂
+      simp only [lineMap_apply, vsub_eq_sub, Pi.vadd_apply', Pi.smul_apply, Pi.sub_apply,
+        smul_eq_mul, vadd_eq_add] at hs₂
+      rw [← neg_sub (w₁ i) (w₂ i), mul_neg, div_mul_cancel₀ _ hp.ne'] at hs₂
+      simp at hs₂
```

Comment:
There seems to be quite a lot of duplication here. Are we missing a useful lemma/abstraction?